### PR TITLE
feat: Added logs_ip_action + more

### LINKED ISSUE #1162: DB struct auth: logs_ip_actions
This is part of: https://github.com/azerothcore/azerothcore-wotlk/issues/1118

Missing table in `auth` db:

```
DROP TABLE IF EXISTS `logs_ip_actions`;
CREATE TABLE IF NOT EXISTS `logs_ip_actions` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT COMMENT 'Unique Identifier',
  `account_id` int(10) unsigned NOT NULL COMMENT 'Account ID',
  `character_guid` int(10) unsigned NOT NULL COMMENT 'Character Guid',
  `type` tinyint(3) unsigned NOT NULL,
  `ip` varchar(15) NOT NULL DEFAULT '127.0.0.1',
  `systemnote` text COMMENT 'Notes inserted by system',
  `unixtime` int(10) unsigned NOT NULL COMMENT 'Unixtime',
  `time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'Timestamp',
  `comment` text COMMENT 'Allows users to add a comment',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COMMENT='Used to log ips of individual actions';
```

### PR DESCRIPTION
**Pull request by Nefertumm**
##### CHANGES PROPOSED:
-  Port from https://github.com/TrinityCore/TrinityCore/commit/6949735098144e478451e73179ca2d9c6e7344f7

###### ISSUES ADDRESSED:

Updates #1118 
Closes #1162 

##### TESTS PERFORMED:

Compiled + basic test (log, create a char, delete a char, try to log with a different password, etc.)

##### HOW TO TEST THE CHANGES:

Modify the conf and play with authserver, like try to log an account that doesn't exist, log with wrong credentials, ban an account and try to log, etc.

##### Target branch(es):

Master

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
what's the status of this? ready for testing?

**Comment by Nefertumm** [comment]
It's ready to be tested since i created the pr ^^

**Comment by wetbrownsauce** [comment]
I haven't played around much with logging prior to this so if someone more knowledgeable wants to take a stab at testing, be my guest. I can report the following:

With logging set to highest:
- Wrong password attempts: log with source IP, Account ID in `logs_ip_actions` as well as logging the accepting connection in `logs`
- Non-existent account attempt: log that a connection was accepted in `logs`
- Banned account connect attempt: log that banned account NAME tried to connect in `logs`

`logs_ip_actions` only seems to be tracking wrong password attempts, not sure what else it does, but the `logs` table records everything above. As I said, not familiar with logging so if you want a specific test or more thorough testing just give me some clear steps and I'll try my best

**Comment by FrancescoBorzi** [comment]
@wetbrownsauce thanks a lot for your feedback, it is very useful.

@Nefertumm what do you think about it?

File: src/server/scripts/ScriptLoader.cpp [hunk 1]
```diff
@@ -648,8 +651,11 @@ void AddWorldScripts()
     AddSC_npcs_special();
     AddSC_npc_taxi();
     AddSC_achievement_scripts();
-    AddSC_chat_log();
+    AddSC_chat_log(); // location: scripts\World\chat_log.cpp
     AddSC_character_creation();
+    // To avoid duplicate code, we check once /*ONLY*/ if logging is permitted or not.
+    if (sWorld->getBoolConfig(CONFIG_IP_BASED_ACTION_LOGGING))
```
Comment (Nefertumm) [inline src/server/scripts/ScriptLoader.cpp]: I don't know why trinity did this, but AddSC_action_ip_logger(); will never be executed, at least in our code...
The loader always loads first. This causes the pr be useless, it doesn't log in logs_ip_action when a character is deleted, created, etc... We should delete the if and place it somewhere else.

**Comment by wetbrownsauce** [comment]
The branch has conflicts now that #1226 has been merged btw

**Comment by Viste** [comment]
@wetbrownsauce no conflicts now )

**Comment by wetbrownsauce** [comment]
@Viste nice, I can test tomorrow!

**Comment by wetbrownsauce** [comment]
I'm still new to logging, so I'm a little confused testing this. What sort of things are supposed to get recorded in `logs_ip_actions` and `logs`? I only see wrong password attempts getting logged in `logs_ip_actions`. `logs` seems to only record that it is accepting connections, whether or not they are locked to an ip/country, and if a banned account tried to log in.  

I'm using in authserver.conf:
```
LogLevel = 3
DebugLogMask = 64
LogFileLevel = 3
EnableLogDB = 1
DBLogLevel = 3
WrongPass.Logging = 1
```

**Comment by FrancescoBorzi** [comment]
@Nefertumm what is the status of this? can it be merged?

**Comment by BarbzYHOOL** [comment]
@wetbrownsauce I think that's pretty much what it is about, talked about this with Nef and he told me it was to log these kind of attempts.

**Comment by FrancescoBorzi** [comment]
feel free to merge then

**Comment by Nefertumm** [comment]
> 
> 
> I'm still new to logging, so I'm a little confused testing this. What sort of things are supposed to get recorded in `logs_ip_actions` and `logs`? I only see wrong password attempts getting logged in `logs_ip_actions`. `logs` seems to only record that it is accepting connections, whether or not they are locked to an ip/country, and if a banned account tried to log in.
> 
> I'm using in authserver.conf:
> 
> ```
> LogLevel = 3
> DebugLogMask = 64
> LogFileLevel = 3
> EnableLogDB = 1
> DBLogLevel = 3
> WrongPass.Logging = 1
> ```

Allow.IP.Based.Action.Logging = 0 -> Set this to 1 to test, otherwise you aren't testing my PR.

**Comment by wetbrownsauce** [comment]
Didn't know there were more options in worldserver.conf related to logging. I'll try it out properly now. 

Why is it marked as to be merged if not properly tested?

**Comment by wetbrownsauce** [comment]
Alright, added in 
`Allow.IP.Based.Action.Logging = 1`
along with a few other logging options. Tried logging into accounts, creating/deleting chars, wrong pass attempts, and banned attempts. Results as follows:

`logs_ip_actions` results:
log

`logs` results:
more_log


Anything further you want me to check into?

**Comment by BarbzYHOOL** [comment]
Awesome, so it's useful when we suspect there are hacked accounts?

**Comment by FrancescoBorzi** [comment]
@wetbrownsauce do you think this can be merged?

**Comment by wetbrownsauce** [comment]
I'm just not sure what all is supposed to be logged since I haven't played with logging until this PR, so I can't tell if anything is missing/not working.

**Comment by FrancescoBorzi** [comment]
if this is not a regression (i.e. something broken that before this PR used to work) then I would merge it,
keeping it pending for too long is not good and can lead to merge conflicts

**Comment by BarbzYHOOL** [comment]
@FrancescoBorzi Plz, merge this if you think it's fine !

**Comment by FrancescoBorzi** [comment]
@Nefertumm @wetbrownsauce good job!

**Pull request merged**